### PR TITLE
DAOS-5912 rebuild: restore pool map if extend/drain fail

### PR DESCRIPTION
**Pull request by wangdi1**
restore the pool map if extend/drain fail.
Add tests to verify it.
    
Test-tag: pr rebuild ec
 Features: rebuild ec
 Required-githooks: true
 Signed-off-by: Di Wang <di.wang@intel.com>

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/common/pool_map.c [hunk 1]
```diff
@@ -1590,8 +1603,12 @@ gen_pool_buf(struct pool_map *map, struct pool_buf **map_buf_out, int map_versio
 
 	rc = add_domain_tree_to_pool_buf(map, map_buf, map_version, dss_tgt_nr, ndomains,
 					 domains, ordered_ranks);
-	if (rc != 0)
+	if (rc != 0) {
+		/* Do not need update the pool map anymore */
+		if (rc == -DER_EXIST)
+			rc = 0;
 		D_GOTO(out_map_buf, rc);
+	}
```
Comment (liw) [inline src/common/pool_map.c:1611]: [Not requesting changes] I was confused during the review that a function named `gen_pool_buf` may return 0 while not generating a map buffer. One could define a positive return value for `gen_pool_buf` to indicate the -DER_EXIST case here, for example, if he is uncomfortable with a common negative return value. Or, at least a comment for `gen_pool_buf` should be added to warn innocent users.

File: src/common/pool_map.c [hunk 2]
```diff
@@ -1551,8 +1553,19 @@ add_domain_tree_to_pool_buf(struct pool_map *map, struct pool_buf *map_buf,
...
+		 * still need extend the pool, though does not need update the pool map
+		 */
+		if (!updated && found_new_dom)
+			return -DER_EXIST;
+
+		/* Those ranks already exists in the pool map, and they are not new, then return
+		 * EALREADY
+		 */
+		if (!updated && !found_new_dom)
+			return -DER_ALREADY;
```
Comment (liw) [inline src/common/pool_map.c:1567]: [Question] Does returning -DER_ALREADY mean that we don't need to update the pool map? I can't find anywhere handling this -DER_ALREADY.

**Review by liw — APPROVED** [review]
I'm not very familiar with the rebuild/extend status change logic.

**Comment by wangdi1** [comment]
failure is due to DAOS-12572.

File: src/common/pool_map.c (hunk 1 above)
Comment (wangdi1) [reply src/common/pool_map.c:1611]: yeah, indeed, export DER_EXIST to the caller might make thing clearer.  I will update it in the following patch, if it is only for better understanding. Thanks.

File: src/common/pool_map.c (hunk 2 above)
Comment (wangdi1) [reply src/common/pool_map.c:1567]: oh, yes, it means the target can not be added. DER_ALREADY will actually return to the user to notify the user all target has been added to the pool map, so "dmg pool extend" failed.

**Review by jolivier23 — APPROVED** [review]

**Pull request merged**